scripts/safeharbor/src/generateUpdates.js

```diff
@@ -0,0 +1,222 @@
+            function: "removeChains",
+            args: [chainIdsToRemove],
+            calldata: agreementInterface.encodeFunctionData("removeChains", [
+                chainIdsToRemove,
+            ]),
+        });
+    }
+
+    // Add new chains from CSV - batch them together
+    if (chainsToAdd.length > 0) {
+        const newChains = chainsToAdd.map((chainName) => {
+            const chainId = getChainId(chainName);
+            const accounts = csvState[chainName] || [];
+
+            return {
+                assetRecoveryAddress: getAssetRecoveryAddress(chainName),
+                accounts: accounts,
+                caip2ChainId: chainId,
+            };
+        });
+
+        // Debug: Check for undefined values in accounts across all new chains
+        newChains.forEach((chain, index) => {
+            if (chain.accounts.length > 0) {
+                const problematicAccounts = chain.accounts.filter(
+                    (acc) =>
+                        !acc.accountAddress ||
+                        acc.childContractScope === undefined ||
+                        acc.childContractScope === null,
+                );
+                if (problematicAccounts.length > 0) {
+                    throw new Error(
+                        `Problematic accounts found in chain ${chainsToAdd[index]}: ${JSON.stringify(problematicAccounts)}`,
+                    );
+                }
+            }
+        });
+
+        updates.push({
+            function: "addChains",
+            args: [newChains],
+            calldata: agreementInterface.encodeFunctionData("addChains", [
+                newChains,
+            ]),
+        });
+    }
+
+    return { updates, chainsToRemove };
+}
+
+function wrapWithMulticall(
+    updates,
+    agreementContractAddress,
+    multicallContractAddress,
+) {
+    // If no updates, return the original array
+    if (updates.length === 0) {
+        return updates;
+    }
+
+    // Convert individual updates to multicall format
+    const calls = updates.map((update) => ({
+        target: agreementContractAddress,
+        callData: update.calldata,
+    }));
+
+    // Generate multicall calldata
+    const multicallCalldata = multicallInterface.encodeFunctionData(
+        "aggregate",
+        [calls],
+    );
+
+    // Add the multicall update to the array
+    const multicallUpdate = {
+        function: "multicall",
+        args: [calls],
+        calldata: multicallCalldata,
+        target: multicallContractAddress,
+    };
+
+    // Return original updates plus the multicall wrapper
+    return [...updates, multicallUpdate];
+}
+
+export function generateUpdates(onChainState, csvState) {
+    const { updates: chainUpdates, chainsToRemove } = generateChainUpdates(
+        onChainState,
+        csvState,
+    );
+    const accountUpdates = generateAccountUpdates(
+        onChainState,
+        csvState,
+        chainsToRemove,
+    );
+    return wrapWithMulticall(
```

Comment:
I believe that this should be 1 level up.
The multicall is not part of the core logic of this module, which is finding the updates required, it's more of a infrastructure/periphery concern.